fuel-asm/src/lib.rs

```diff
@@ -1,26 +1,578 @@
-//! FuelVM opcodes representation
+//! FuelVM instruction and opcodes representation.
 
 #![cfg_attr(docsrs, feature(doc_auto_cfg))]
 #![cfg_attr(not(feature = "std"), no_std)]
 #![cfg_attr(feature = "std", doc = include_str!("../README.md"))]
 #![warn(missing_docs)]
 
-mod instruction;
-mod opcode;
+mod gm_args;
+mod instruction_result;
+// This is `pub` to make documentation for the private `impl_instructions!` macro more accessible.
+#[macro_use]
+pub mod macros;
+pub mod op;
+mod pack;
 mod panic_reason;
+mod unpack;
 
-mod macros;
-
+pub use fuel_types::{RegisterId, Word};
 #[doc(no_inline)]
-pub use fuel_types::{self, Immediate06, Immediate12, Immediate18, Immediate24, RegisterId, Word};
-pub use instruction::Instruction;
-pub use opcode::{GMArgs, GTFArgs, Opcode, OpcodeRepr};
-pub use panic_reason::{InstructionResult, PanicReason};
+pub use gm_args::{GMArgs, GTFArgs};
+pub use instruction_result::InstructionResult;
+pub use panic_reason::PanicReason;
 
-#[cfg(feature = "serde")]
-#[doc(no_inline)]
-pub use serde;
+/// Represents a 6-bit register ID, guaranteed to be masked by construction.
+#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct RegId(u8);
```

Comment:
These `RegId` and `Imm*` types are probably better off in the `fuel-types` crate, potentially replacing the old `RegisterId` and `Immediate*` types as a type-safe alternative? Doing so would require a lot more breakage throughout fuel-vm and downstream, so I thought I'd leave that for follow-up work as this PR's already quite big.